keyserver/listener.go

```diff
@@ -0,0 +1,108 @@
+package keyserver
+
+import (
+	"crypto/tls"
+	"io"
+	"log"
+	"net"
+	"time"
+
+	. "github.com/coniks-sys/coniks-go/protocol"
+)
+
+func (server *ConiksServer) listenForRequests(ln *net.TCPListener, handler func(msg []byte) ([]byte, error)) {
+	defer ln.Close()
+	for {
+		select {
+		case <-server.stop:
+			server.waitCloseConn.Wait()
+			return
+		}
+		ln.SetDeadline(time.Now().Add(5 * time.Second))
+		conn, err := ln.Accept()
+		if err != nil {
+			if opErr, ok := err.(*net.OpError); ok && opErr.Timeout() {
+				continue
+			}
+			log.Printf("accept client: %s", err)
+			continue
+		}
+		conn = tls.Server(conn, server.tlsConfig)
+		server.waitCloseConn.Add(1)
+		go func() {
+			server.acceptClient(conn, handler)
+			server.waitCloseConn.Done()
+		}()
+	}
+}
+
+func (server *ConiksServer) acceptClient(conn net.Conn, handler func(msg []byte) ([]byte, error)) {
+	defer conn.Close()
+	conn.SetDeadline(time.Now().Add(5 * time.Second))
+
+	// handle request
+	buf := make([]byte, 4096)
+	for {
+		n, err := conn.Read(buf)
+		if err != nil {
+			if err != io.EOF {
+				log.Printf("client read %v: %v", conn.RemoteAddr(), err)
+			}
+			return
+		}
+
+		res, err := handler(buf[:n])
+		if err != nil {
+			log.Printf("client handle %v: %v", conn.RemoteAddr(), err)
+		}
+
+		n, err = conn.Write([]byte(res))
+		if err != nil {
+			log.Printf("client write %v: %v", conn.RemoteAddr(), err)
+			return
+		}
+	}
+}
+
+func malformedClientMsg(err error) ([]byte, error) {
+	// check if we're just propagating a message
+	if err == nil {
+		err = ErrorMalformedClientMessage.Error()
+	}
+	response := NewErrorResponse(ErrorMalformedClientMessage)
+	res, e := MarshalResponse(response)
+	if e != nil {
+		panic(e)
+	}
+	return res, err
+}
+
+func (server *ConiksServer) makeHandler(acceptableTypes map[int]bool) func(msg []byte) ([]byte, error) {
+	return func(msg []byte) ([]byte, error) {
+		// get request message
+		req, err := UnmarshalRequest(msg)
+		if err != nil {
+			return malformedClientMsg(err)
+		}
+		if !acceptableTypes[req.Type] {
+			log.Printf("unacceptable message type: %q", req.Type)
+			return malformedClientMsg(ErrorMalformedClientMessage.Error())
+		}
+
+		switch req.Type {
+		default:
+			server.Lock()
+		}
+		response, e := server.dir.HandleOps(req)
+		switch req.Type {
+		default:
+			server.Unlock()
+		}
+
+		res, err := MarshalResponse(response)
+		if err != nil {
+			panic(e)
```

Comment:
I think the panic should be on `err` not `e`, no? Otherwise, if the registration request resulted in an error as well, we'll end up panicking on that instead.